Tools/plot_parallel.py

```diff
@@ -158,95 +163,64 @@ def get_field_max( filename, a_field ):
-        plot_snapshot( filename )
-        if plot_Ey_max_evolution:
-            my_zwin[count], my_maxF[count] = get_field_max( filename, 'Ey' )
-
-    if plot_Ey_max_evolution:
-        # Gather window position and field max arrays to rank 0
-        zwin_rbuf = None
-        maxF_rbuf = None
-        if rank == 0:
-            zwin_rbuf = np.empty([size, max_buf_size], dtype='d')
-            maxF_rbuf = np.empty([size, max_buf_size], dtype='d')
-        comm_world.Gather(my_zwin, zwin_rbuf, root=0)
-        comm_world.Gather(my_maxF, maxF_rbuf, root=0)
-        # Re-format 2D arrays zwin_rbuf and maxF_rbuf on rank 0
-        # into 1D arrays, and plot them
-        if rank == 0:
-            zwin_arr = np.zeros( nfiles )
-            maxF_arr = np.zeros( nfiles )
-            istart = 0
-            for i in range(size):
-                nelem = nfiles_list[i]
-                zwin_arr[istart:istart+nelem] = zwin_rbuf[i,0:nelem]
-                maxF_arr[istart:istart+nelem] = maxF_rbuf[i,0:nelem]
-                istart += nelem
-            # Plot evolution of field max
-            plot_field_max()
-else:
-    ### Serial analysis ###
-    print('Serial analysis')
-    print('Number of plotfiles: %s' %nfiles)
-    pslist = get_species(file_list);
+    try:
+        from mpi4py import MPI
+        comm_world = MPI.COMM_WORLD
+        rank = comm_world.Get_rank()
+        size = comm_world.Get_size()
+    except ImportError:
+        pass
+
+if rank == 0:
+    print('number of MPI ranks: %d'%size)
+    print('Number of plotfiles: %s'%nfiles)
     print('list of species: ', pslist)
-    if plot_Ey_max_evolution:
-        zwin_arr = np.zeros( nfiles )
-        maxF_arr = np.zeros( nfiles )
-    # Loop over files and
-    # - plot field snapshot
-    # - store window position and field max in arrays
-    for count, filename in enumerate(file_list):
-        plot_snapshot( filename )
-        if plot_Ey_max_evolution:
-            zwin_arr[count], maxF_arr[count] = get_field_max( filename, 'Ey' )
-    # Plot evolution of field max
-    if plot_Ey_max_evolution:
-        plot_field_max()
+
+if plot_max_evolution is not None:
+    # Fill with a value less than any possible value
+    zwin = np.full(nfiles, np.finfo(float).min)
+    maxF = np.full(nfiles, np.finfo(float).min)
+
+# Loop over files, splitting plotfile list among MPI ranks
+# - plot field snapshot
+# - store window position and field max in arrays
+for count, filename in enumerate(file_list):
+    if count%size != rank:
+        continue
+
+    plot_snapshot( filename )
+    if plot_max_evolution is not None:
+        zwin[count], maxF[count] = get_field_max( filename, plot_max_evolution )
+
+if plot_max_evolution is not None:
+    if size > 1:
+        global_zwin = np.empty_like(zwin)
+        global_maxF = np.empty_like(maxF)
+        comm_world.Reduce(zwin, global_zwin, op=MPI.MAX)
+        comm_world.Reduce(maxF, global_maxF, op=MPI.MAX)
```

Comment:
Ok my bad, this does give the expected result (sorry I don't get what the Reduce function does).

Reply:
For each element in the array passed in, reduce finds the maximum value of the element over the processors.

Reply:
Oh I get it now, it is actually an element-wise `Reduce` operation, and for a given index all ranks have `np.finfo(float).min` except the rank that operated on the index (which has a >0 value).

Then this would break if the user is interested in a quantity that can be <0. If a user needs the average transverse position of the beam (which I am actually doing now), all negative values will be replaced by zeros. Could we add a safeguard for this?

Reply:
np.finfo(float).min is the lowest, most negative, float, typically -1.7976931348623157e+308.